circuit-retained/src/commonMain/kotlin/com/slack/circuit/retained/RememberRetained.kt

```diff
@@ -16,68 +27,34 @@ import androidx.compose.runtime.remember
+ * Navigators. The same caveats of `rememberSaveable` also still apply (i.e. do not retain Android
+ * Contexts, Views, etc).
  *
- * However, it does not participate in saved instance state either, so care should be taken to
- * choose the right retention mechanism for your use case. Consider the below two examples.
+ * By specifying a [saver], you can also serialize the retained value to saved instance state, in
+ * addition to keeping it in memory, to persist some state through process death.
  *
- * The first case will retain `state` across configuration changes but will _not_ survive process
- * death.
- *
- * ```kotlin
- * @Composable
- * override fun present(): CounterState {
- *   var state by rememberRetained { mutableStateOf(CounterState(0)) }
- *
- *   return CounterState(count) { event ->
- *     when (event) {
- *       is CounterEvent.Increment -> state = state.copy(count = state.count + 1)
- *       is CounterEvent.Decrement -> state = state.copy(count = state.count - 1)
- *     }
- *   }
- * }
- * ```
- *
- * This second case will retain `count` across configuration changes _and_ survive process death.
- * However, it only works with primitives or `Parcelable` state types.
- *
- * ```kotlin
- * @Composable
- * override fun present(): CounterState {
- *   var count by rememberSaveable { mutableStateOf(0) }
- *
- *   return CounterState(count) { event ->
- *     when (event) {
- *       is CounterEvent.Increment -> state = count++
- *       is CounterEvent.Decrement -> state = count--
- *     }
- *   }
- * }
- * ```
+ * Care should be taken to choose the right retention mechanism for your use case.
  *
  * @param inputs A set of inputs such that, when any of them have changed, will cause the state to
  *   reset and [init] to be rerun
+ * @param saver The [Saver] object which defines how the state is saved and restored.
  * @param key An optional key to be used as a key for the saved value. If not provided we use the
  *   automatically generated by the Compose runtime which is unique for the every exact code
  *   location in the composition tree
  * @param init A factory function to create the initial value of this state
  */
 @Composable
-public fun <T : Any> rememberRetained(vararg inputs: Any?, key: String? = null, init: () -> T): T {
-  val registry = LocalRetainedStateRegistry.current
-  // Short-circuit no-ops
-  if (registry === NoOpRetainedStateRegistry) {
-    return when (inputs.size) {
-      0 -> remember(init)
-      1 -> remember(inputs[0], init)
-      2 -> remember(inputs[0], inputs[1], init)
-      3 -> remember(inputs[0], inputs[1], inputs[2], init)
-      else -> remember(keys = inputs, init)
-    }
-  }
+public fun <T : Any> rememberRetained(
+  vararg inputs: Any?,
+  saver: Saver<T, out Any> = Saver({ null }, { null }),
+  key: String? = null,
+  init: () -> T,
+): T {
+  val saveableStateRegistry = LocalSaveableStateRegistry.current
```

Comment:
I think this would be possible if the underlying implementation is updated to use a private `saver: Saver?` trick where `null` distinguishes between the two overloads. If we made the `saver: Saver?` argument public in an overload though, that seems like it would be trickier, since it'd be possible (but weird) to pass in a `null` `saver` the first time, and then later swap it out for a non-null one.

Reply:
Talked offline, that sounds good 👍

Reply:
I tinkered a bunch with this locally and could never arrive at an impl that wasn't weird and unwieldy, so I think we're ok to just keep this merged impl. Then we just either make `saver` explicit or offer a `rememberRetainedSaveable` that's semantically explicit and uses `autoSaver()`